### ISSUE TITLE
[feature] "set xxx" message for catch~ and send~

### ISSUE THREAD

**Issue by jamshark70** [pure-data/pure-data#1567]
I'll lead off by saying that I did look at the source and I understand why [catch~] and [send~] don't implement `set` (because [throw~] and [receive~] rely on `pd_findbyclass()` and this enforces one and only one object defining a specific name).

But I have a use case where, given the available interfaces, I can't think of any good way to do it.

I would like to make modular mixer objects, following roughly this design.

![pd-mixer-design](https://user-images.githubusercontent.com/318301/151906806-a5a1e2d0-44bc-4e34-8bf6-b1e726434b0e.png)

If I use a mixer's name for signal routing objects, then, the immediate problem. I'd have two choices to let the user edit the name:

- Use a symbol box for user entry (populated initially by an argument). User-friendly UI, but it's a nonstarter, because there is no message that I can send to [catch~] to give it a new symbol.

- Or, the user could edit the object string and re-create the abstraction instance. This means all of the instance's local state gets lost. 1/ User creates a mixer with name 'abc'. 2/ User carefully sets level, panning etc. 3/ User decides, no, 'abc' isn't descriptive, let me call it 'lead'. 4/ At this point, all sliders jump back to 0. User curses.

I'm considering a workaround of storing mixer parameters in a global [text] -- however, here it would be very helpful to have a singleton (#1370) so that the user would not have to create the global table manually, but this doesn't exist either.

Or, I could generate throw~ / catch~ names internally independent of the user-facing labels, but I would still need a global table correlating the labels to the routing names.

Or I could give up and require everything to be patched by hand (but I was really hoping for a clean UI). Or, give up and make it non-modular.

Or maybe all of this is a bad idea (though that would be an unfortunate conclusion).

To sum up, I'm stuck: If I want to avoid a global table, then I need to be able to `set xyz` [catch~] and [send~] objects. If that isn't going to happen, then I need a global table (and I don't know how to make this transparent to the end-user).

**Comment by Spacechild1** [comment]
>  If you change a [catch~] object's name, you need to update the x_whereto on any throw~ that matches either the old or new names: 

See step 3. above.

**Comment by jamshark70** [comment]
> See step 3. above.

Yes. My question is about HOW to do this in C, lol. (I'm spoiled by SC: You can just 'do' over a collection.)

Anyway I guess you can do a PR faster than I can.

**Comment by Spacechild1** [comment]
> My question is about HOW to do this in C, lol. (I'm spoiled by SC: You can just 'do' over a collection.)

`canvas_updatedsp()`.

Don't bother with the PR, though, unless you feel like doing it.

For the meantime, I just made a nice little abstraction called `[dynamic-catch~]`: [dynamic-catch~.zip](https://github.com/pure-data/pure-data/files/7979630/dynamic-catch.zip)

The abstractions needs `iemguts`. It can be done in Pd vanilla, but it would be rather awkward.

**Comment by Spacechild1** [comment]
Here's a corresponding `[dynamic-send~]` object: [dynamic-send~.zip](https://github.com/pure-data/pure-data/files/7979761/dynamic-send.zip)

Dynamically named send/receive pairs can already be emulated with `[tabsend~]` + `[tabreceive~]`. The former needs a corresponding array that can be renamed with the `[rename(` method.

One problem you will notice with both `[receive~]` and `[tabreceive~]` is that they will complain if there's no matching `[send~]` resp. array. It would be nice to have an option to disable this... Note that `[throw~]` and `[tabsend~]` do *not* complain if there's no matching `[catch~]` resp. array.

**Comment by Spacechild1** [comment]
As a side note, you can also implement "busses" in Pd like this:

![grafik](https://user-images.githubusercontent.com/16126632/152003033-7808232a-9308-4d53-b151-d75c86cfc07a.png)

Everybody can write to the bus with `[throw~ <name>]` and everyone can receive from the bus with `[r~ <name>]`. This might be useful for your mixer project.

**Comment by jamshark70** [comment]
> For the meantime, I just made a nice little abstraction called [dynamic-catch~]: dynamic-catch~.zip

Oh, these look great, thanks!

> It can be done in Pd vanilla, but it would be much more difficult. You can try if you're looking for a challenge :-)

I'm not, thanks :wink: 

> Dynamically named send/receive pairs can already be emulated with [tabsend~] + [tabreceive~]. The former needs a corresponding array that can be renamed with the [rename( method.

Hm ... links among help files ... Couldn't ~throw~-help.pd~ say "If you need a [catch~] whose name can be changed dynamically, see tabsend-receive~-help.pd"? I can understand with the minimalist core philosophy that it might be considered redundant to have both catch~ and tabreceive~ being dynamically reassignable, but in that case, it would be helpful to point the user to the other, related objects.

**Comment by Spacechild1** [comment]
> "If you need a [catch~] whose name can be changed dynamically, see tabsend-receive~-help.pd"?

Just to be clear: `[tabsend~]` + `[tabreceive~]` is a replacement for `[send~]` + `[receive~]`, but *not* for `[throw~]` + `[catch~]`.

**Comment by umlaeute** [comment]
> As a side note, you can also implement "busses" in Pd like this:

that looks bogus.
it should be

~~~
[catch~ $1]
|
[send~ $1]
~~~

(unless you want things to really explode)

**Comment by Spacechild1** [comment]
> that looks bogus.

Oh my, thanks for catching~ that!

**Comment by jamshark70** [comment]
> Just to be clear: [tabsend~] + [tabreceive~] is a replacement for [send~] + [receive~], but not for [throw~] + [catch~].

Oh, I see, [tabreceive~] won't mix.

But... the whole design may be a nonstarter anyway, if [throw~] and [catch~] always add a block of delay. In SC, I can arrange mixer channels in the order source --> target, and mixers incur no latency. In Pd, it looks like the only way to force order is using an explicit patch cable (which I was hoping to avoid).

So I guess I have to reconsider the use case. At the same time, the original proposed feature would be nice.

**Comment by umlaeute** [comment]
> if [throw~] and [catch~] always add a block of delay.

they don't (at least not *always*)
